### PR TITLE
All-in-one Certmanager helm chart

### PR DESCRIPTION
**Pull request by chrissng**
TODO:
- [x] solver for GCP CloudDNS
- [x] solver for Amazon Route53
- [ ] ?? sync cert secrets in multiple namespaces (https://cert-manager.netlify.com/docs/faq/kubed/)

### CODE REVIEW

File: cert-manager-aio/values.yaml [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+acme_solver: clouddns
+
+route53:
+  aws_region: ap-southeast-1
+  aws_access_key_id: ""
+  aws_secret_access_key: ""
+
+clouddns:
+  gcp_project_id: ""
+  service_account_key: ""
```
Comment (lawliet89) [inline cert-manager-aio/values.yaml]: Possible for this and the AWS ones to be left `null` to use the instance metadata server?

Comment (chrissng) [reply cert-manager-aio/values.yaml]: Done

Comment (chrissng) [reply cert-manager-aio/values.yaml]: Done

File: cert-manager-aio/templates/secret.yaml [hunk 2]
```diff
@@ -0,0 +1,16 @@
+---
```
Comment (lawliet89) [inline cert-manager-aio/templates/secret.yaml]: Don't render this file at all if either credentials are omitted.

Comment (chrissng) [reply cert-manager-aio/templates/secret.yaml]: Fixed

File: cert-manager-aio/templates/certificate.yaml [hunk 3]
```diff
@@ -0,0 +1,14 @@
+---
```
Comment (lawliet89) [inline cert-manager-aio/templates/certificate.yaml]: Can this be made optional?

Comment (lawliet89) [inline cert-manager-aio/templates/certificate.yaml]: Any way to make this be repeatable? Like allow the user to specify multiple certificates.

File: cert-manager-aio/values.yaml [hunk 4]
```diff
@@ -26,4 +26,5 @@ clouddns:
   # Set service_account_key to null if getting credentials via metadata server
   service_account_key: null
 
-domain_name: example.com
+domain_names:
```
Comment (lawliet89) [inline cert-manager-aio/values.yaml]: Something more like:

```yaml
domain_names:
  - cn: xxx
    san:
      - xxx
      - xxx
```

Comment (chrissng) [reply cert-manager-aio/values.yaml]: Good suggestion, i have expanded this to allow specifying when to renew before expiry also.

File: cert-manager-aio/templates/certificate.yaml (hunk 3 above)
Comment (chrissng) [reply cert-manager-aio/templates/certificate.yaml]: Allowed specifying as lists, so empty means no certs are created.

File: cert-manager-aio/templates/certificate.yaml [hunk 5]
```diff
@@ -0,0 +1,19 @@
...
+  name: {{ include "cert-manager-aio.fullname" $root }}-cert-{{ $i }}
+spec:
+  secretName: {{ include "cert-manager-aio.fullname" $root }}-cert-{{ $i }}
+  issuerRef:
+    name: {{ include "cert-manager-aio.fullname" $root }}-issuer
+    kind: ClusterIssuer
+  renewBefore: {{ $domain.renew_before }}
+  commonName: {{ $domain.common_name | quote }}
+  dnsNames:
+    {{- with $domain.san }}
```
Comment (lawliet89) [inline cert-manager-aio/templates/certificate.yaml]: Not sure if certmanager does it, but the CN must always be included in the SAN too.

Comment (chrissng) [reply cert-manager-aio/templates/certificate.yaml]: In case it doesn't, the chart will include the common name in the san if not in it already.

**Pull request merged**